### PR TITLE
Sfc tests

### PR DESCRIPTION
**Pull request by devintegral4**
tests for sfc v1.1 contract

### CODE REVIEW

File: contracts/sfc/Staker.sol [hunk 1]
```diff
@@ -511,6 +523,34 @@ contract Stakers is Ownable, StakersConstants {
...
+        require(stakers[stakerID].deactivatedTime == 0, "staker is deactivated");
+        require(amount != 0, "zero amount");
+
+        // don't allow to withdraw full as a request, because amount==0 originally meant "not existing"
+        uint256 totalAmount = stakers[stakerID].stakeAmount;
+        require(amount + minStake() <= totalAmount, "must leave at least minStake");
+        uint256 newAmount = totalAmount - amount;
+        require((newAmount.mul(maxDelegatedRatio())).div(RATIO_UNIT) >= stakers[stakerID].delegatedMe, "too much delegations");
+
+        require(withdrawalRequests[staker][wrID].amount == 0, "wrID already exists");
```
Comment (quan8) [inline contracts/sfc/Staker.sol]: move this check near the beginning, say after line 529?

Comment (quan8) [reply contracts/sfc/Staker.sol]: withdrawalRequests[staker][wrID]:  add new variable?

File: contracts/sfc/Staker.sol [hunk 2]
```diff
@@ -522,11 +562,15 @@ contract Stakers is Ownable, StakersConstants {
 
         uint256 stake = stakers[stakerID].stakeAmount;
         uint256 penalty = 0;
-        bool isCheater = stakers[stakerID].status & CHEATER_MASK != 0;
+        uint256 status = stakers[stakerID].status;
```
Comment (quan8) [inline contracts/sfc/Staker.sol]: add some logs into these kinds of methods, if possible

File: contracts/sfc/Staker.sol [hunk 3]
```diff
@@ -562,31 +607,109 @@ contract Stakers is Ownable, StakersConstants {
         emit DeactivatedDelegation(from, stakerID);
     }
 
-    event WithdrawnDelegation(address indexed from, uint256 indexed stakerID, uint256 penalty);
+    function prepareToWithdrawDelegationPartial(uint256 wrID, uint256 amount) external {
+        address payable delegator = msg.sender;
```
Comment (quan8) [inline contracts/sfc/Staker.sol]: add logs

File: contracts/sfc/Staker.sol [hunk 4]
```diff
@@ -562,31 +607,109 @@ contract Stakers is Ownable, StakersConstants {
...
+        emit WithdrawnDelegation(delegator, stakerID, penalty);
+    }
+
+    event WithdrawnByRequest(address indexed auth, address indexed receiver, uint256 indexed stakerID, uint256 wrID, uint256 penalty);
+
+    function withdrawByRequest(uint256 wrID) external {
+        address auth = msg.sender;
+        address payable receiver = msg.sender;
+        require(withdrawalRequests[auth][wrID].time != 0, "request doesn't exist");
+        bool delegation = withdrawalRequests[auth][wrID].delegation;
```
Comment (quan8) [inline contracts/sfc/Staker.sol]: may introduce variable for withdrawalRequests[auth][wrID]

File: contracts/sfc/Staker.sol [hunk 5]
```diff
@@ -108,7 +108,8 @@ contract Stakers is Ownable, StakersConstants {
 
         uint256 delegatedMe;
 
-        address stakerAddress;
+        address dagAddress;
```
Comment (quan8) [inline contracts/sfc/Staker.sol]: add comment. Difference of dagAddress and sfcAddress?

Comment (quan8) [reply contracts/sfc/Staker.sol]: can dagAddress = sfcAddress?

File: contracts/sfc/Staker.sol [hunk 6]
```diff
@@ -250,39 +252,98 @@ contract Stakers is Ownable, StakersConstants {
     Methods
     */
 
-    event CreatedStake(uint256 indexed stakerID, address indexed stakerAddress, uint256 amount);
+    event CreatedStake(uint256 indexed stakerID, address indexed dagSfcAddress, uint256 amount);
 
     // Create new staker
     // Stake amount is msg.value
+    // dagAdrress is msg.sender
```
Comment (quan8) [inline contracts/sfc/Staker.sol]: dagAdrress spelling => dagAddress
Fix all occurrences in the file

File: contracts/sfc/Staker.sol [hunk 7]
```diff
@@ -250,39 +252,98 @@ contract Stakers is Ownable, StakersConstants {
...
+        }
+        if (stakers[stakerID].sfcAddress != sfcAddress) {
+            return 0;
+        }
+        return stakerID;
+    }
+
+    event UpdatedStakerSfcAddress(uint256 indexed stakerID, address indexed oldSfcAddress, address indexed newSfcAddress);
+
+    function updateStakerSfcAddress(address newSfcAddress) external {
```
Comment (quan8) [inline contracts/sfc/Staker.sol]: add method's description

File: test/Stakers.js [hunk 8]
```diff
@@ -13,13 +13,14 @@ contract('Staker test', async ([firstStaker, secondStaker, thirdStaker, firstDep
     beforeEach(async () => {
         this.firstEpoch = 0;
         this.stakers = await UnitTestStakers.new(this.firstEpoch);
+        this.validatorComission = new BN('150000'); // 0.15
```
Comment (quan8) [inline test/Stakers.js]: Comission spelling => Commission

File: contracts/sfc/Staker.sol [hunk 9]
```diff
@@ -53,19 +53,19 @@ contract StakersConstants {
     }
 
     function unbondingStartDate() public pure returns (uint256) {
-      return 1577419000;
+        return 1577419000;
```
Comment (quan8) [inline contracts/sfc/Staker.sol]: add comment, which date?

File: contracts/sfc/Staker.sol [hunk 10]
```diff
@@ -543,20 +543,20 @@ contract Stakers is Ownable, StakersConstants {
...
         address payable delegator = msg.sender;
 
         require(delegations[delegator].amount != 0, "delegation doesn't exist");
         require(delegations[delegator].deactivatedTime == 0, "delegation is deactivated");
-        (uint256 pendingRewards, uint256 fromEpoch, uint256 untilEpoch) = calcDelegationRewards(delegator, _fromEpoch, maxEpochs);
+        (uint256 pendingRewards, uint256 fromEpoch, uint256 untilEpoch) = calcDelegationRewards(delegator, 0, maxEpochs);
 
         require(delegations[delegator].paidUntilEpoch < fromEpoch, "epoch is already paid");
         require(fromEpoch <= currentSealedEpoch, "future epoch");
         require(untilEpoch >= fromEpoch, "no epochs claimed");
```
Comment (quan8) [inline contracts/sfc/Staker.sol]: fromEpoch <= untilEpoch : is more readable.

File: contracts/sfc/Staker.sol [hunk 11]
```diff
@@ -543,20 +543,20 @@ contract Stakers is Ownable, StakersConstants {
     event ClaimedDelegationReward(address indexed from, uint256 indexed stakerID, uint256 reward, uint256 fromEpoch, uint256 untilEpoch);
 
     // Claim the pending rewards for a given delegator (sender)
-    // _fromEpoch is starting epoch which rewards are calculated (including). If 0, then it's lowest not claimed epoch
     // maxEpochs is maximum number of epoch to calc rewards for. Set it to your chunk size.
-    function claimDelegationRewards(uint256 _fromEpoch, uint256 maxEpochs) external {
+    function claimDelegationRewards(uint256 maxEpochs) external {
```
Comment (quan8) [inline contracts/sfc/Staker.sol]: can claim rewards partially?

File: contracts/sfc/Staker.sol [hunk 12]
```diff
@@ -566,23 +566,23 @@ contract Stakers is Ownable, StakersConstants {
     event ClaimedValidatorReward(uint256 indexed stakerID, uint256 reward, uint256 fromEpoch, uint256 untilEpoch);
 
     // Claim the pending rewards for a given stakerID (sender)
-    // _fromEpoch is starting epoch which rewards are calculated (including). If 0, then it's lowest not claimed epoch
     // maxEpochs is maximum number of epoch to calc rewards for. Set it to your chunk size.
     //
     // may be already deactivated, but still allowed to withdraw old rewards
-    function claimValidatorRewards(uint256 _fromEpoch, uint256 maxEpochs) external {
+    function claimValidatorRewards(uint256 maxEpochs) external {
```
Comment (quan8) [inline contracts/sfc/Staker.sol]: can claim rewards partially?

File: contracts/sfc/Staker.sol [hunk 13]
```diff
@@ -775,9 +775,12 @@ contract Stakers is Ownable, StakersConstants {
     function withdrawDelegation() external {
         address payable delegator = msg.sender;
         require(delegations[delegator].deactivatedTime != 0, "delegation wasn't deactivated");
-        require(block.timestamp >= delegations[delegator].deactivatedTime + delegationLockPeriodTime(), "not enough time passed");
-        require(currentEpoch() >= delegations[delegator].deactivatedEpoch + delegationLockPeriodEpochs(), "not enough epochs passed");
         uint256 stakerID = delegations[delegator].toStakerID;
+        if (stakers[stakerID].stakeAmount != 0) {
```
Comment (quan8) [inline contracts/sfc/Staker.sol]: - add method description

Comment (quan8) [reply contracts/sfc/Staker.sol]: line 780, perhaps: "// check if withdrawal of delegation is permitted"

File: contracts/sfc/Staker.sol [hunk 14]
```diff
@@ -20,6 +20,10 @@ contract StakersConstants {
         return 1 * 1e18;
```
Comment (quan8) [inline contracts/sfc/Staker.sol]: can make a constant for 1e18?

File: test/Stakers.js [hunk 15]
```diff
@@ -397,13 +397,17 @@ contract('Staker test', async ([firstStaker, secondStaker, thirdStaker, firstDep
...
 
         // check early withdrawal
+        await expectRevert(this.stakers.prepareToWithdrawStake({from: firstStaker}), 'not all rewards claimed');
+        await this.stakers.discardValidatorRewards({from: firstStaker});
         await this.stakers.prepareToWithdrawStake({from: firstStaker}); // deactivate staker
         {
             time.increase(86400 * 7);
             await this.stakers._makeEpochSnapshots(10000);
-            await this.stakers._makeEpochSnapshots(10000);
+            await this.stakers._makeEpochSnapshots(10000;
```
Comment (quan8) [inline test/Stakers.js]: missing )

File: contracts/sfc/Staker.sol [hunk 16]
```diff
@@ -609,6 +609,25 @@ contract Stakers is Ownable, StakersConstants {
...
     }
 
+    // stashed rewards are burnt on deactivation in all the cases except when delegator has deactivated after
+    // validator has deactivated or was slashed/pruned
+    function _rewardsBurnableOnDeactivation(bool isDelegation, uint256 stakerID) public view returns(bool) {
+        return !isDelegation || (stakers[stakerID].stakeAmount != 0 && stakers[stakerID].status == OK_STATUS && stakers[stakerID].deactivatedTime == 0);
+    }
+
+    // proportional part of stashed rewards are burnt on deactivation if _rewardsBurnableOnDeactivation returns true
+    function _mayBurnRewardsOnDeactivation(bool isDelegation, uint256 stakerID, address addr, uint256 withdrawAmount, uint256 totalAmount) internal {
```
Comment (quan8) [inline contracts/sfc/Staker.sol]: - perhaps can improve readability by removing isDelegation parameter in both methods _rewardsBurnableOnDeactivation and _mayBurnRewardsOnDeactivation.

=>
function _rewardsBurnableOnDeactivation(uint256 stakerID) public view returns(bool) {
        return (stakers[stakerID].stakeAmount != 0 && stakers[stakerID].status == OK_STATUS && stakers[stakerID].deactivatedTime == 0);
    }


function _burnRewardsOnDeactivation (bool isDelegation, uint256 stakerID, address addr, uint256 withdrawAmount, uint256 totalAmount) internal { 
   uint256 leftAmount = totalAmount.sub(withdrawAmount);
            uint256 newStash = rewardsStash[addr].amount.mul(leftAmount).div(totalAmount);
            if (newStash == 0) {
                delete rewardsStash[addr];
            } else {
                rewardsStash[addr].amount = newStash;
            }
}

Then line 641 (similarly line 666)
_mayBurnRewardsOnDeactivation(false, stakerID, stakerSfcAddr, stakers[stakerID].stakeAmount, stakers[stakerID].stakeAmount);
=>
// add comment: isDelegation is false
_burnRewardsOnDeactivation(stakerID, stakerSfcAddr, stakers[stakerID].stakeAmount, stakers[stakerID].stakeAmount);


Then line 725, 754:
_mayBurnRewardsOnDeactivation(true, stakerID, delegator, delegations[delegator].amount, delegations[delegator].amount);
=>
// add comment why isDelegation is true
if (_rewardsBurnableOnDeactivation(stakerID)) {
    _burnRewardsOnDeactivation(true, stakerID, delegator, delegations[delegator].amount, delegations[delegator].amount);
}

File: contracts/sfc/Staker.sol (hunk 12 above)
Comment (devintegral3) [reply contracts/sfc/Staker.sol]: Of course, number of non-claimed epochs may be too great to calculate all rewards in one call

**Comment by devintegral3** [comment]
Hi @quan8! Really appreciated your comments. This PR adds tests for the new SFC features, and mostly doesn't change the contract code. Please see PRs 14-28

**Pull request closed without merging**